README.md
# color-converter

## 介绍

该扩展可以帮助你直接转换色值，支持 `hex` => `rgb` => `hsl` => `hex` 转换，支持多个颜色选择转换。

## 使用方式

选择一个或者按住 `Alt` 选择多个颜色，例如

- `hex`
  - `#f44034b8`
  - `##ff0fdb`
  - `#f1d`
- `rgb`
  - `rgba(244, 64, 52, 0.72)`
  - `rgb(244, 64, 52)`
  - `rgb(255, 15, 219)`
- `hsl`
  - `hsla(4, 90%, 58%, 0.72)`
  - `hsl(4, 90%, 58%)`
  - `hsl(309, 100%, 53%)`

### 方式一（推荐）

直接使用 `Ctrl + Alt + R`（mac 上为 `cmd + Alt + R`） 即可切换颜色。

### 方式二

`Alt + Shift + P` 呼出控制台，输入 `cc` 即可转换颜色。

## 依赖

该扩展无需其他依赖。

## 扩展配置

该扩展暂无配置。

## 已知问题

1. 选择了色值却无法转换，请注意可能是选择中前后有空格，因为正则中对于一个有效的色值判断为 `/^[#rh].+\)$/`（并不准确，请不要学我），所以如果前后有空格则不生效。

## 后续更新

1. 对于色值的正则判断需要改改。
2. 是否需要增加一个扩展配置，配置转换后的 `rgb` 或者 `hsl` 中是否有空格。
3. `rgba(0,0,0,.7)` 无法正常处理，并且 `rgba(0,0,0,0.7)` 之后透明度消失。

## 版本

### 0.1.0

初版提交，初步功能完成。

### 0.1.2

将转换后过长的位数给处理掉了，如果有小数的话会保留一位小数。

### 0.1.5

修复 `rgba(0,0,0,.7)` 无法正常处理的问题，修复 `rgba(0,0,0,0.7)` 透明度消失的问题。


CHANGELOG.md
# Change Log

All notable changes to the "color-converter" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/Color.js
/*
 * @Author: jsjzh
 * @Email: kimimi_king@163.com
 * @Date: 2019-05-27 15:48:38
 * @LastEditors: jsjzh
 * @LastEditTime: 2019-05-29 10:26:29
 * @Description: Color 类
 * hex -> rgb -> hsl -> hex
 */

const {
  testColor,
  removeSpace,
  transBracketColor,
  hex2rgb,
  rgb2rgb,
  hsl2rgb,
  rgb2hex,
  rgb2hsl,
  formatRgb,
  formatHsl,
  formatHex
} = require('./utils')

const vscode = require('vscode')

const color2rgb = { hex2rgb, rgb2rgb, hsl2rgb }

class Color {
  constructor(color) {
    color = removeSpace(color)

    const { type } = testColor.find(regObj => regObj.reg.test(color)) || {}

    this.initColor = color
    this.initType = type

    this._r = 255
    this._g = 255
    this._b = 255
    this._a = 1
    this.invalid = true

    if (this.initType) {
      this.invalid = false
      this.init()
    }
  }

  init() {
    let color = this.initColor
    switch (this.initType) {
      case 'hex':
        color = [color]
        break
      case 'rgb':
        color = transBracketColor(color)
        break
      case 'hsl':
        color = transBracketColor(color)
        break
    }
    const { r, g, b, a = 1 } = color2rgb[`${this.initType}2rgb`].apply(null, color)
    // 在此处控制 rgba 四个值都为 number
    this._r = +r
    this._g = +g
    this._b = +b
    this._a = +a
  }

  getColor() {
    let { _r: r, _g: g, _b: b, _a: a } = this
    let result
    switch (this.initType) {
      case 'hex':
        result = formatRgb(rgb2rgb(r, g, b, a))
        break
      case 'rgb':
        result = formatHsl(rgb2hsl(r, g, b, a))
        break
      case 'hsl':
        result = formatHex(rgb2hex(r, g, b, a))
        break
    }
    return result
  }
}

module.exports = Color


src/index.js
const vscode = require('vscode')
const { Range, Position } = vscode

const Color = require('./Color')
const { testColor } = require('./utils')

exports.activate = context => {
  console.log('color converter is activate')

  function colorConverter() {
    const { activeTextEditor } = vscode.window
    const { selections, document } = activeTextEditor
    // selections，去除空的以及非单行的 selections
    const preSelections = selections.filter(selection => !selection.isEmpty && selection.isSingleLine)
    // positions
    const prePositions = preSelections.map(selection => [
      new Position(selection._start._line, selection._start._character),
      new Position(selection._end._line, selection._end._character)
    ])
    // Ranges
    const preRanges = prePositions.map(validPositions => new Range(validPositions[0], validPositions[1]))
    // 获取有效的 Ranges，该数组中的为用户选中并且通过了色值 reg 的过滤的 ranges
    const validRanges = preRanges.filter(range => testColor.some(regObj => regObj.reg.test(document.getText(range))))
    // 存储每个色值的类
    const colors = validRanges.map(range => new Color(document.getText(range)))
    // 实际实现修改的方法
    activeTextEditor.edit(editBuilder => {
      validRanges.forEach((range, index) => {
        editBuilder.replace(range, colors[index].getColor())
      })
    })
  }

  const realCommand = vscode.commands.registerCommand('extension.cc', colorConverter)

  context.subscriptions.push(realCommand)
}

exports.deactivate = () => {
  console.log('color converter is deactivate')
}


temp/colorsys.js
const RGB_MAX = 255
const HUE_MAX = 360
const SV_MAX = 100

var colorsys = (module.exports = {})

colorsys.rgb2Hsl = function(r, g, b) {
  if (typeof r === 'object') {
    const args = r
    r = args.r
    g = args.g
    b = args.b
  }
  // It converts [0,255] format, to [0,1]
  r = r === RGB_MAX ? 1 : (r % RGB_MAX) / parseFloat(RGB_MAX)
  g = g === RGB_MAX ? 1 : (g % RGB_MAX) / parseFloat(RGB_MAX)
  b = b === RGB_MAX ? 1 : (b % RGB_MAX) / parseFloat(RGB_MAX)

  var max = Math.max(r, g, b)
  var min = Math.min(r, g, b)
  var h,
    s,
    l = (max + min) / 2

  if (max === min) {
    h = s = 0 // achromatic
  } else {
    var d = max - min
    s = l > 0.5 ? d / (2 - max - min) : d / (max + min)
    switch (max) {
      case r:
        h = (g - b) / d + (g < b ? 6 : 0)
        break
      case g:
        h = (b - r) / d + 2
        break
      case b:
        h = (r - g) / d + 4
        break
    }
    h /= 6
  }

  return {
    h: Math.floor(h * HUE_MAX),
    s: Math.floor(s * SV_MAX),
    l: Math.floor(l * SV_MAX)
  }
}

colorsys.rgb_to_hsl = colorsys.rgbToHsl = colorsys.rgb2Hsl

colorsys.rgb2Hsv = function(r, g, b) {
  if (typeof r === 'object') {
    const args = r
    r = args.r
    g = args.g
    b = args.b
  }

  // It converts [0,255] format, to [0,1]
  r = r === RGB_MAX ? 1 : (r % RGB_MAX) / parseFloat(RGB_MAX)
  g = g === RGB_MAX ? 1 : (g % RGB_MAX) / parseFloat(RGB_MAX)
  b = b === RGB_MAX ? 1 : (b % RGB_MAX) / parseFloat(RGB_MAX)

  var max = Math.max(r, g, b)
  var min = Math.min(r, g, b)
  var h,
    s,
    v = max

  var d = max - min

  s = max === 0 ? 0 : d / max

  if (max === min) {
    h = 0 // achromatic
  } else {
    switch (max) {
      case r:
        h = (g - b) / d + (g < b ? 6 : 0)
        break
      case g:
        h = (b - r) / d + 2
        break
      case b:
        h = (r - g) / d + 4
        break
    }
    h /= 6
  }

  return {
    h: Math.floor(h * HUE_MAX),
    s: Math.floor(s * SV_MAX),
    v: Math.floor(v * SV_MAX)
  }
}

colorsys.rgb_to_hsv = colorsys.rgbToHsv = colorsys.rgb2Hsv

colorsys.hsl2Rgb = function(h, s, l) {
  if (typeof h === 'object') {
    const args = h
    h = args.h
    s = args.s
    l = args.l
  }

  var r, g, b

  h = h === HUE_MAX ? 1 : (h % HUE_MAX) / parseFloat(HUE_MAX)
  s = s === SV_MAX ? 1 : (s % SV_MAX) / parseFloat(SV_MAX)
  l = l === SV_MAX ? 1 : (l % SV_MAX) / parseFloat(SV_MAX)

  if (s === 0) {
    r = g = b = l // achromatic
  } else {
    var hue2rgb = function hue2rgb(p, q, t) {
      if (t < 0) t += 1
      if (t > 1) t -= 1
      if (t < 1 / 6) return p + (q - p) * 6 * t
      if (t < 1 / 2) return q
      if (t < 2 / 3) return p + (q - p) * (2 / 3 - t) * 6
      return p
    }

    var q = l < 0.5 ? l * (1 + s) : l + s - l * s
    var p = 2 * l - q
    r = hue2rgb(p, q, h + 1 / 3)
    g = hue2rgb(p, q, h)
    b = hue2rgb(p, q, h - 1 / 3)
  }
  return { r: Math.round(r * RGB_MAX), g: Math.round(g * RGB_MAX), b: Math.round(b * RGB_MAX) }
}

colorsys.hsl_to_rgb = colorsys.hslToRgb = colorsys.hsl2Rgb

colorsys.hsv2Rgb = function(h, s, v) {
  if (typeof h === 'object') {
    const args = h
    h = args.h
    s = args.s
    v = args.v
  }

  h = h === HUE_MAX ? 1 : ((h % HUE_MAX) / parseFloat(HUE_MAX)) * 6
  s = s === SV_MAX ? 1 : (s % SV_MAX) / parseFloat(SV_MAX)
  v = v === SV_MAX ? 1 : (v % SV_MAX) / parseFloat(SV_MAX)

  var i = Math.floor(h)
  var f = h - i
  var p = v * (1 - s)
  var q = v * (1 - f * s)
  var t = v * (1 - (1 - f) * s)
  var mod = i % 6
  var r = [v, q, p, p, t, v][mod]
  var g = [t, v, v, q, p, p][mod]
  var b = [p, p, t, v, v, q][mod]

  return { r: Math.round(r * RGB_MAX), g: Math.round(g * RGB_MAX), b: Math.round(b * RGB_MAX) }
}

colorsys.hsv_to_rgb = colorsys.hsv2Rgb
colorsys.hsvToRgb = colorsys.hsv2Rgb

colorsys.rgb2Hex = function(r, g, b) {
  if (typeof r === 'object') {
    const args = r
    r = args.r
    g = args.g
    b = args.b
  }
  r = Math.round(r).toString(16)
  g = Math.round(g).toString(16)
  b = Math.round(b).toString(16)

  r = r.length === 1 ? '0' + r : r
  g = g.length === 1 ? '0' + g : g
  b = b.length === 1 ? '0' + b : b

  return '#' + r + g + b
}

colorsys.rgb_to_hex = colorsys.rgbToHex = colorsys.rgb2Hex

colorsys.hex2Rgb = function(hex) {
  var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex)
  return result
    ? {
        r: parseInt(result[1], 16),
        g: parseInt(result[2], 16),
        b: parseInt(result[3], 16)
      }
    : null
}

colorsys.hex_to_rgb = colorsys.hexToRgb = colorsys.hex2Rgb

colorsys.hsv2Hex = function(h, s, v) {
  var rgb = colorsys.hsv2Rgb(h, s, v)
  return colorsys.rgb2Hex(rgb.r, rgb.g, rgb.b)
}

colorsys.hsv_to_hex = colorsys.hsv2Hex
colorsys.hsvToHex = colorsys.hsv2Hex

colorsys.hex2Hsv = function(hex) {
  var rgb = colorsys.hex2Rgb(hex)
  return colorsys.rgb2Hsv(rgb.r, rgb.g, rgb.b)[0]
}

colorsys.hex_to_hsv = colorsys.hexToHsv = colorsys.hex2Hsv

colorsys.hsl2Hex = function(h, s, l) {
  var rgb = colorsys.hsl2Rgb(h, s, l)
  return colorsys.rgb2Hex(rgb.r, rgb.g, rgb.b)
}

colorsys.hsl_to_hex = colorsys.hslToHex = colorsys.hsl2Hex

colorsys.hex2Hsl = function(hex) {
  var rgb = colorsys.hex2Rgb(hex)
  return colorsys.rgb2Hsl(rgb.r, rgb.g, rgb.b)[0]
}

colorsys.hex_to_hsl = colorsys.hexToHsl = colorsys.hex2Hsl

colorsys.rgb2cmyk = function(r, g, b) {
  if (typeof r === 'object') {
    const args = r
    r = args.r
    g = args.g
    b = args.b
  }

  var rprim = r / 255
  var gprim = g / 255
  var bprim = b / 255

  var k = 1 - Math.max(rprim, gprim, bprim)

  var c = (1 - rprim - k) / (1 - k)
  var m = (1 - gprim - k) / (1 - k)
  var y = (1 - bprim - k) / (1 - k)

  return {
    c: c.toFixed(3),
    m: m.toFixed(3),
    y: y.toFixed(3),
    k: k.toFixed(3)
  }
}

colorsys.rgb_to_cmyk = colorsys.rgbToCmyk = colorsys.rgb2Cmyk

colorsys.cmyk2rgb = function(c, m, y, k) {
  if (typeof c === 'object') {
    const args = c
    c = args.c
    m = args.m
    y = args.y
    k = args.k
  }

  var r = 255 * (1 - c) * (1 - k)
  var g = 255 * (1 - m) * (1 - k)
  var b = 255 * (1 - y) * (1 - k)

  return {
    r: Math.floor(r),
    g: Math.floor(g),
    b: Math.floor(b)
  }
}

colorsys.cmyk_to_rgb = colorsys.cmykToRgb = colorsys.cmyk2Rgb

colorsys.hsv2Hsl = function(h, s, v) {
  if (typeof h === 'object') {
    const args = h
    h = args.h
    s = args.s
    v = args.v
  }

  var l = ((2 - s) * v) / 2

  if (l !== 0) {
    if (l === SV_MAX) {
      s = 0
    } else if (l < SV_MAX / 2) {
      s = (s * v) / (l * 2)
    } else {
      s = (s * v) / (2 - l * 2)
    }
  }

  return { h, s, l }
}

colorsys.hsv_to_hsl = colorsys.hsvToHsl = colorsys.hsv2Hsl

colorsys.hsl2Hsv = function(h, s, l) {
  if (typeof h === 'object') {
    const args = h
    h = args.h
    s = args.s
    l = args.l
  }

  s = s * (l < 50 ? l : 100 - l)

  return {
    h: h,
    s: Math.floor((2 * s) / (l + s)),
    v: Math.floor(l + s)
  }
}

colorsys.hsl_to_hsv = colorsys.hslToHsv = colorsys.hsl2Hsv


src/utils.js
/*
 * @Author: jsjzh
 * @Email: kimimi_king@163.com
 * @Date: 2019-05-27 15:48:38
 * @LastEditors: jsjzh
 * @LastEditTime: 2019-05-29 14:15:58
 * @Description: 工具函数集
 */
const RGB_MAX = 255
const HUE_MAX = 360
const SV_MAX = 100
/**
 * #22b1b180
 * #22b1b1
 * #fff
 */
const HexReg = /^\#[A-Fa-f0-9]{3,8}$/
/**
 * rgba(34,177,177,0.5)
 * rgb(34,177,177)
 */
const rgbReg = /^(rgb|rgba){1}\([\d,. ]+\)$/
/**
 * hsla(180,68%,41%,0.5)
 * hsl(180,68%,41%)
 */
const hslReg = /^(hsl|hsla){1}\([\d,%. ]+\)$/

const testColor = [{ type: 'hex', reg: HexReg }, { type: 'rgb', reg: rgbReg }, { type: 'hsl', reg: hslReg }]

/**
 * 去除空格
 * @param {String} str
 */
function removeSpace(str) {
  return str.replace(' ', '')
}

/**
 * 获取 xxx(a(%)?,b(%)?,c(%)?,d(%)?,e(%)?,...) 中 abcde... 的值
 * 并且，在此会控制 abcde... 等等的值都为 number
 * @param {String} str
 */
function transBracketColor(str) {
  return str
    .replace(/([rgbahsl()%])+?/g, '')
    .split(',')
    .map(num => +num)
}

/**
 * 输入 hex 转换为 rgb，将所有的输入都统一成 #xxxxxxxx
 * @param {String} str #xxx #xxxxxx #xxxxxxxx
 */
function hex2rgb(str) {
  let reg = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i
  str = str.charAt(0) === '#' ? str.slice(1) : str
  if (str.length === 3) {
    let _r = str.charAt(0)
    let _g = str.charAt(1)
    let _b = str.charAt(2)
    str = _r + _r + _g + _g + _b + _b
  }
  if (str.length === 6) {
    str = str + 'ff'
  }
  let result = reg.exec(str)

  return {
    r: parseInt(result[1], 16),
    g: parseInt(result[2], 16),
    b: parseInt(result[3], 16),
    a: parseInt(result[4], 16) / 255
  }
}

/**
 * hsl 转 rgb，同样的，所有的输出都有 a（透明度），在输出的时候处理 a
 * @param {Number} h [0,360]
 * @param {Number} s [0,100]
 * @param {Number} l [0,100]
 * @param {Number} a [0,1]
 */
function hsl2rgb(h, s, l, a = 1) {
  let r, g, b

  h = h === HUE_MAX ? 1 : (h % HUE_MAX) / parseFloat(HUE_MAX)
  s = s === SV_MAX ? 1 : (s % SV_MAX) / parseFloat(SV_MAX)
  l = l === SV_MAX ? 1 : (l % SV_MAX) / parseFloat(SV_MAX)

  if (s === 0) {
    r = g = b = l
  } else {
    function hue2rgb(p, q, t) {
      if (t < 0) t += 1
      if (t > 1) t -= 1
      if (t < 1 / 6) return p + (q - p) * 6 * t
      if (t < 1 / 2) return q
      if (t < 2 / 3) return p + (q - p) * (2 / 3 - t) * 6
      return p
    }

    let q = l < 0.5 ? l * (1 + s) : l + s - l * s
    let p = 2 * l - q
    r = hue2rgb(p, q, h + 1 / 3)
    g = hue2rgb(p, q, h)
    b = hue2rgb(p, q, h - 1 / 3)
  }

  return { r: r * RGB_MAX, g: g * RGB_MAX, b: b * RGB_MAX, a }
}

/**
 * rgb 转 hex，还是老样子，全部转为完整的格式，即使透明度是 1
 * @param {Number} r [0,255]
 * @param {Number} g [0,255]
 * @param {Number} b [0,255]
 * @param {Number} a [0,1]
 */
function rgb2hex(r, g, b, a = 1) {
  r = Math.round(r).toString(16)
  g = Math.round(g).toString(16)
  b = Math.round(b).toString(16)
  a = Math.round(a * 255).toString(16)

  r = r.length === 1 ? '0' + r : r
  g = g.length === 1 ? '0' + g : g
  b = b.length === 1 ? '0' + b : b
  a = a.length === 1 ? '0' + a : a

  return '#' + r + g + b + a
}

/**
 * rgb 转 hsl，全部转为完整的格式，即使透明度为 1
 * @param {Number} r [0,255]
 * @param {Number} g [0,255]
 * @param {Number} b [0,255]
 * @param {Number} a [0,1]
 */
function rgb2hsl(r, g, b, a = 1) {
  r = r === RGB_MAX ? 1 : (r % RGB_MAX) / parseFloat(RGB_MAX)
  g = g === RGB_MAX ? 1 : (g % RGB_MAX) / parseFloat(RGB_MAX)
  b = b === RGB_MAX ? 1 : (b % RGB_MAX) / parseFloat(RGB_MAX)

  let max = Math.max(r, g, b)
  let min = Math.min(r, g, b)
  let h,
    s,
    l = (max + min) / 2

  if (max === min) {
    h = s = 0
  } else {
    let d = max - min
    s = l > 0.5 ? d / (2 - max - min) : d / (max + min)
    switch (max) {
      case r:
        h = (g - b) / d + (g < b ? 6 : 0)
        break
      case g:
        h = (b - r) / d + 2
        break
      case b:
        h = (r - g) / d + 4
        break
    }
    h /= 6
  }

  return { h: h * HUE_MAX, s: s * SV_MAX, l: l * SV_MAX, a }
}

/**
 * 输入即输出函数，不做转换
 * @param {Number} r [0,255]
 * @param {Number} g [0,255]
 * @param {Number} b [0,255]
 * @param {Number} a [0,1]
 */
function rgb2rgb(r, g, b, a = 1) {
  return { r, g, b, a }
}

/**
 * 数字保留小数专用
 * @param {Number} num 2.2222
 * @param {Number} point 1
 */
function roundPoint(num, point = 1) {
  let multiple = Math.pow(10, point)
  return Math.round(num * multiple) / multiple
}

/**
 * 格式化 rgb，若 a === 1，返回 rgb，否则都返回 rgba
 * @param {Number} r [0,255]
 * @param {Number} g [0,255]
 * @param {Number} b [0,255]
 * @param {Number} a [0,1]
 */
function formatRgb({ r, g, b, a = 1 }) {
  a = +a
  r = roundPoint(r)
  g = roundPoint(g)
  b = roundPoint(b)
  a = roundPoint(a)
  return a === 1 ? `rgb(${r}, ${g}, ${b})` : `rgba(${r}, ${g}, ${b}, ${a})`
}

/**
 * 格式化 hsl，若 a === 1，返回 hsl，否则都返回 hsla
 * @param {Number} h [0,360]
 * @param {Number} s [0,100]
 * @param {Number} l [0,100]
 * @param {Number} a [0,1]
 */
function formatHsl({ h, s, l, a = 1 }) {
  a = +a
  h = roundPoint(h)
  s = roundPoint(s)
  l = roundPoint(l)
  a = roundPoint(a)
  return a === 1 ? `hsl(${h}, ${s}%, ${l}%)` : `hsl(${h}, ${s}%, ${l}%, ${a})`
}

/**
 * 输入完整的（八位）hex 色值，若透明度为 1，返回六位 hex，若还可精简则返回三位 hex
 * @param {String} hex #xxxxxxxx
 */
function formatHex(hex) {
  hex = hex.toLocaleLowerCase()
  hex = hex.endsWith('ff') ? hex.slice(0, -2) : hex
  if (
    hex.charAt(1) === hex.charAt(2) &&
    hex.charAt(3) === hex.charAt(4) &&
    hex.charAt(5) === hex.charAt(6) &&
    (!hex.charAt(7) && !hex.charAt(8))
  ) {
    hex = '#' + hex.charAt(1) + hex.charAt(3) + hex.charAt(5)
  }
  return hex
}

module.exports = {
  testColor,
  removeSpace,
  transBracketColor,
  hex2rgb,
  rgb2rgb,
  hsl2rgb,
  rgb2hex,
  rgb2hsl,
  formatRgb,
  formatHsl,
  formatHex
}
